src/libcollections/str.rs

```diff
@@ -1634,30 +1634,53 @@ impl str {
     #[stable(feature = "unicode_case_mapping", since = "1.2.0")]
     pub fn to_lowercase(&self) -> String {
         let mut s = String::with_capacity(self.len());
-        for (i, c) in self[..].char_indices() {
-            if c == 'Σ' {
-                // Σ maps to σ, except at the end of a word where it maps to ς.
-                // This is the only conditional (contextual) but language-independent mapping
-                // in `SpecialCasing.txt`,
-                // so hard-code it rather than have a generic "condition" mechanism.
-                // See https://github.com/rust-lang/rust/issues/26035
-                map_uppercase_sigma(self, i, &mut s)
-            } else {
-                s.extend(c.to_lowercase());
+        let mut left = None;
+
+        // Try to collect slices of lower case characters to push into the
+        // result or extend with the lower case version if an upper case
+        // character is found.
+        for (i, ch) in self.char_indices() {
+            if !ch.is_lowercase() && ch.is_alphabetic() {
```

Comment:
> From what I understand non-alphabetic characters can't be turned to lower case,

Unicode is full of surprising corner cases, so I wouldn’t rely on something like this on faith.

I’m opposed to skipping `char::to_lowercase` unless you make `src/etc/unicode.py` check exhaustively that it returns its input unchanged for ever code point where you would skip it, and that this stays the case when we updated to a new Unicode version.

> The issue is it would defeat the purpose of the optimization, because the conversion is what takes most of the time.

Here are a few random ideas to improved this.

`char::to_lowercase` is implemented in `src/librustc_unicode/tables.rs` with a binary search, while `char::is_lowercase` and `char::is_alphabetic` use a trie of bits *and* have a fast-path for ASCII.

`char::to_lowercase` could use a similar `BoolTrie` to skip the binary search for code points that are unchanged (thought this would have a cost in binary size) and have its own ASCII fast path.

`to_lowercase_table` could be changed to store `[u8; 9]` (or whatever is the maximum length) in UTF-8 rather than `[char; 3]` in (effectively) UTF-32, so that pushing to a `String` doesn’t need to do an encoding conversion.

If doing larger copies from `&self` turns out to be significantly more efficient than pushing code points one (or up to three) at a time, the `ToLowercase` iterator could be extended to implement `PartialEq<char>` or equivalent, to find out if the conversion was a no-op.